### ISSUE TITLE
Multithreading support?

### ISSUE THREAD

**Issue by Adonai** [proot-me/PRoot#111]
Hi. As I can see PRoot is now only single-threaded and looks at tracee signals sequentially one after another (event.c). Are there any plans on doing handler-per-tracee-thread implementation?

**Comment by jopasserat** [comment]
This is not part of the short-term plans. But feel free to document something over here if you have ideas to address this.

All the efforts are directed at solving #106  as far as I'm concerned.

Fancy joining?

**Comment by Adonai** [comment]
Thanks!

I've yet to understand codebase and implications of changes, but will post here in case of any ideas.

**Comment by jopasserat** [comment]
I've yet to understanding everything as well :)
But I'll try to format some notes that I've gathered here and there and make them available on the wiki so that everyone can contribute their bits too.

**Comment by vincenthage** [comment]
Hi Adonai,
I've found a mention of your idea at the end of the [roadmap](https://github.com/proot-me/PRoot/blob/master/doc/proot/roadmap.txt). Though implementing it reliably might take some time.

It might imply some issues with extensions/features that rely on having one unique central PRoot instance (like the netcoop and copy-on-write extensions).
But performance wise it might be worth it still.

**Comment by oxr463** [comment]
See also: https://github.com/proot-me/proot/issues/106#issuecomment-494680705

**Comment by travarin** [comment]
Hi, is this still something you're interested in adding to the project? My partner and I from UT Austin would be interested in working on this for one of our classes.

**Comment by oxr463** [comment]
Yes please! Also, if you have any experience with seccomp, then please help!

See: https://github.com/proot-me/proot/pull/130

**Comment by travarin** [comment]
Just to clarify, for libseccomp would we be implementing the tests under the TODO in #130?

**Comment by oxr463** [comment]
You mean these?

> Create tests for seccomp:
> - with PROOT_NO_SECCOMP
> - without PROOT_NO_SECCOMP
> - kernel version >= 4.8
> - kernel version < 4.8

If so, the answer is yes; we need to be sure that PRoot will work properly regardless of kernel version or seccomp configuration.
